pkg/container-collection/options.go

```diff
@@ -34,77 +31,128 @@ import (
 	"github.com/kinvolk/inspektor-gadget/pkg/runcfanotify"
 
 	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
 	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/k8s"
 	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/pubsub"
 )
 
-// WithDockerEnrichment automatically adds container metadata with Docker
+func containerRuntimeEnricher(
+	runtimeName string,
+	runtimeClient runtimeclient.ContainerRuntimeClient,
+	container *pb.ContainerDefinition,
+) bool {
+	// Is container already enriched?
+	if container.Name != "" && container.Namespace != "" && container.Podname != "" {
+		return true
+	}
+
+	containers, err := runtimeClient.GetContainers(container.Id)
+	if err != nil {
+		log.Debugf("Runtime enricher (%s): failed to get container %q: %s",
+			runtimeName, container.Id, err)
+		return true
+	}
+
+	switch len(containers) {
+	case 0:
+		log.Debugf("Runtime enricher (%s): none container found with ID %q",
+			runtimeName, container.Id)
+	case 1:
+		name := strings.TrimPrefix(containers[0].Name, "/")
+		if len(name) == 0 {
+			log.Debugf("Runtime enricher (%s): empty name for container %q",
+				runtimeName, container.Id)
+			break
+		}
+
+		// Notice that, at this point, the container was already
+		// enriched with the PID by the hook.
+		container.Name = name
+
+		// Some gadgets require the namespace and pod name to be set
+		container.Namespace = "default"
+		container.Podname = container.Name
+	default:
+		log.Errorf("Runtime enricher (%s): container ID %q has multiple (%d) containers. Drop it",
+			runtimeName, container.Id, len(containers))
+
+		// Drop container
+		return false
+	}
+
+	// Do not drop container. Let other runtimes try to enrich it.
+	return true
+}
+
+// WithContainerRuntimeEnrichment automatically adds the container name using
+// the CRI (for CRI-O and Containerd) or the Docker Engine API.
+//
+// Notice that it also sets the container namespace to "default" and the podname
+// equal to the container name. It is done because some gadgets need those two
+// values to be set.
 //
-// ContainerCollection.ContainerCollectionInitialize(WithDockerEnrichment())
-func WithDockerEnrichment() ContainerCollectionOption {
+// ContainerCollection.ContainerCollectionInitialize(WithContainerRuntimeEnrichment(*RuntimeConfig))
+func WithContainerRuntimeEnrichment(runtime *containerutils.RuntimeConfig) ContainerCollectionOption {
 	return func(cc *ContainerCollection) error {
-		dockercli, err := dockerclient.NewClientWithOpts(dockerclient.FromEnv)
+		runtimeClient, err := containerutils.NewContainerRuntimeClient(runtime)
 		if err != nil {
-			return err
+			log.Warnf("Runtime enricher (%s): failed to initialize container runtime",
+				runtime.Name)
+			return nil
 		}
 
-		// Already running containers
-		containers, err := dockercli.ContainerList(context.Background(),
-			dockertypes.ContainerListOptions{
-				All: true,
-			})
+		containers, err := runtimeClient.GetContainers("")
```

Comment:
I'd prefer to have two different functions:
- GetContainers(): no parameters, returns all containers.
- GetContainer(id string): returns only container with given ID